src/main/java/org/openstreetmap/atlas/utilities/caching/strategies/NamespaceCachingStrategy.java

```diff
@@ -0,0 +1,223 @@
+    private static final String FILE_EXTENSION_DOT = ".";
+    private static final String PROPERTY_LOCAL_TEMPORARY_DIRECTORY = "java.io.tmpdir";
+    private static final String TEMPORARY_DIRECTORY_STRING = System
+            .getProperty(PROPERTY_LOCAL_TEMPORARY_DIRECTORY);
+
+    private final String namespace;
+
+    public NamespaceCachingStrategy(final String namespace)
+    {
+        super();
+        if (namespace.contains("/") || namespace.contains("\\"))
+        {
+            throw new IllegalArgumentException(
+                    "The namespace cannot contain characters \'\\\' or \'/\'");
+        }
+        this.namespace = namespace;
+    }
+
+    @Override
+    public Optional<Resource> attemptFetch(final URI resourceURI,
+            final Function<URI, Optional<Resource>> defaultFetcher)
+    {
+        if (TEMPORARY_DIRECTORY_STRING == null)
+        {
+            logger.error("StrategyID {}: failed to read property {}, skipping cache fetch...",
+                    this.getStrategyID(), PROPERTY_LOCAL_TEMPORARY_DIRECTORY);
+            return Optional.empty();
+        }
+
+        if (resourceURI == null)
+        {
+            logger.warn("StrategyID {}: resourceURI was null, skipping cache fetch...",
+                    this.getStrategyID());
+            return Optional.empty();
+        }
+
+        final File cachedFile = getCachedFile(resourceURI);
+        attemptToCacheFileLocally(cachedFile, defaultFetcher, resourceURI);
+
+        if (cachedFile.exists())
+        {
+            logger.trace("StrategyID {}: returning local copy of resource {}", this.getStrategyID(),
+                    resourceURI);
+            return Optional.of(cachedFile);
+        }
+
+        // If we got here, something went wrong in attemptToCacheFileLocally().
+        logger.warn("StrategyID {}: could not find local copy of resource {}", this.getStrategyID(),
+                resourceURI);
+        return Optional.empty();
+    }
+
+    @Override
+    public String getName()
+    {
+        return "NamespaceCachingStrategy";
+    }
+
+    @Override
+    public void invalidate()
+    {
+        final Path storageDirectory = this.getStorageDirectory();
+        try
+        {
+            new File(storageDirectory.toString()).deleteRecursively();
+        }
+        catch (final Exception exception)
+        {
+            logger.warn("StrategyID {}: invalidate failed due to {}", this.getStrategyID(),
+                    exception);
+        }
+    }
+
+    @Override
+    public void invalidate(final URI resourceURI)
+    {
+        try
+        {
+            getCachedFile(resourceURI).delete();
+        }
+        catch (final Exception exception)
+        {
+            logger.warn("StrategyID {}: invalidate of resource {} failed due to {}",
+                    this.getStrategyID(), resourceURI, exception);
+        }
+    }
+
+    private void attemptToCacheFileLocally(final File cachedFile,
```

Comment:
Any point this method should throw an exception instead of just logging errors?

Reply:
I actually don't think it should, since an exception would needlessly halt execution. If `attemptToCacheFileLocally` fails, then what ultimately happens is that the `NamespaceCachingStrategy.attemptFetch(URI)` returns an empty `Optional`. From there, the `ConcurrentResourceCache` object containing that `NamespaceCachingStrategy` will simply fall back to the default fetcher.

I am open to having my mind changed on this.